source/common/quic/http_datagram_handler.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "envoy/http/codec.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "absl/strings/string_view.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Quic {
+
+// Enables HTTP Datagrams and the Capsule Protocol support based on RFC 9297. This class is used
+// both on the receive side (HTTP/3 codec -> Envoy) and the send side (Envoy -> HTTP/3 codec) of
+// the QUIC client and server streams in Envoy.
+class HttpDatagramHandler : public quic::QuicSpdyStream::Http3DatagramVisitor,
+                            public quiche::CapsuleParser::Visitor,
+                            protected Logger::Loggable<Logger::Id::quic_stream> {
+public:
+  // Does not take ownership of the QuicSpdyStream. "stream" must refer to a valid object
+  // that outlives HttpDatagramHandler.
+  explicit HttpDatagramHandler(quic::QuicSpdyStream* stream);
+  ~HttpDatagramHandler() override;
+
+  // quic::QuicSpdyStream::Http3DatagramVisitor
+  // Normalizes HTTP/3 Datagrams to Capsules for Envoy to do subsequent processing.
+  void OnHttp3Datagram(quic::QuicStreamId stream_id, absl::string_view payload) override;
+  void OnUnknownCapsule(quic::QuicStreamId stream_id,
+                        const quiche::UnknownCapsule& capsule) override;
+
+  // quiche::CapsuleParser::Visitor
+  // Converts a given Capsule to a HTTP/3 Datagrams and send it.
+  bool OnCapsule(const quiche::Capsule& capsule) override;
+  void OnCapsuleParseFailure(absl::string_view error_message) override;
+
+  // Re-encodes given Capsule fragment into an HTTP/3 Datagram. Returns true if an HTTP/3 Datagram
+  // created and sent successfully. Returns false if the CapsuleParser fails to parse the
+  // |capsule_fragment| or the corresponding stream fails to send the Datagram.
+  bool encodeCapsuleFragment(absl::string_view capsule_fragment, bool end_stream);
+  // Does not take ownership of the StreamDecoder. "decoder" must outlive HttpDatagramHandler.
+  void setStreamDecoder(Http::StreamDecoder* decoder) { stream_decoder_ = decoder; }
```

Comment:
nit: Consider using a ref instead of a pointer.

Reply:
This is a little tricky to use a reference instead because stream_decoder_ is set after the object initialization. Also the Envoy QUIC stream holds a pointer to a decoder anyway so I would leave it as is unless you have a good alternative in mind.